packages/odyssey-react-mui/src/DataTable/DataTable.tsx

```diff
@@ -0,0 +1,757 @@
+import { DataTableEmptyState } from "./DataTableEmptyState";
+import { Callout } from "../Callout";
+import { t } from "i18next";
+
+export type DataTableGetDataType = {
+  page?: number;
+  resultsPerPage?: number;
+  search?: string;
+  filters?: DataFilter[];
+  sort?: MRT_SortingState;
+};
+
+export type DataTableOnReorderRowsType = {
+  rowId: string;
+  newRowIndex: number;
+};
+
+export type DataTableProps = {
+  /**
+   * The columns that make up the table
+   */
+  columns: MRT_TableOptions<MRT_RowData>["columns"];
+  /**
+   * The total number of rows in the table. Optional, because it's sometimes impossible
+   * to calculate. Used in table pagination to know when to disable the "next"/"more" button.
+   */
+  totalRows?: number;
+  /**
+   * The function to get the ID of a row
+   */
+  getRowId?: MRT_TableOptions<MRT_RowData>["getRowId"];
+  /**
+   * The initial density of the table. This is available even if the table density
+   * isn't changeable.
+   */
+  initialDensity?: (typeof densityValues)[number];
+  /**
+   * If true, the end user will be able to change the table density.
+   */
+  hasChangeableDensity?: boolean;
+  /**
+   * If true, the end user can resize individual columns.
+   */
+  hasColumnResizing?: boolean;
+  /**
+   * If true, the end user will be able to show/hide columns.
+   */
+  hasColumnVisibility?: boolean;
+  /**
+   * If true, the end user will be able to filter columns.
+   */
+  hasFilters?: boolean;
+  /**
+   * If true, the table will include pagination controls.
+   */
+  hasPagination?: boolean;
+  /**
+   * If true, the table will include checkboxes on each row, enabling
+   * the user to select some or all rows.
+   */
+  hasRowSelection?: boolean;
+  /**
+   * If true, the global table search controls will be shown.
+   */
+  hasSearch?: boolean;
+  /**
+   * If true, the end user can sort columns (ascending, descending, or neither)
+   */
+  hasSorting?: boolean;
+  /**
+   * If true, the end user can reorder rows via a drag-and-drop interface
+   */
+  hasRowReordering?: boolean;
+  /**
+   * If true, the search field will include a Search button, rather than
+   * firing on input change.
+   */
+  hasSearchSubmitButton?: boolean;
+  /**
+   * The debounce time, in milliseconds, for the search input firing
+   * `onChangeSearch` when changed. If `hasSearchSubmitButton` is true,
+   * this doesn't do anything.
+   */
+  searchDelayTime?: number;
+  /**
+   * Callback that fires when a row (or rows) is selected or unselected.
+   */
+  onChangeRowSelection?: (rowSelection: DataTableRowSelectionState) => void;
+  /**
+   * Callback that fires whenever the table needs to fetch new data, due to changes in
+   * page, results per page, search input, filters, or sorting
+   */
+  getData: ({
+    page,
+    resultsPerPage,
+    search,
+    filters,
+    sort,
+  }: DataTableGetDataType) =>
+    | MRT_TableOptions<MRT_RowData>["data"]
+    | Promise<MRT_TableOptions<MRT_RowData>["data"]>;
```

Comment:
I feel like we should just stick with this returning a promise? It makes the type much simpler.

Reply:
This needs to be both because the getData function can be either sync or async